### PR TITLE
[REG-1525] Initial doc changes for states and actions

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
Doc updates for State changes in [RGUnityBots PR](https://github.com/Regression-Games/RGUnityBots/pull/131)

If you're wondering where the section is to tell you how to attach things to game objects... 

That is the very change we're making here.  You don't have to do that at all anymore.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: docs/integrating-with-unity/defining-states.mdx [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+# Defining States
+
+:::info
+TODO: We need to write a better description about why it is important to limit the state exposed to your bot.
+If we can't justify this vs accessing game object state directly, we should question what value we add...
+:::
+
+The state of the game sent to Regression Games is limited to only GameObjects selected to be included as part of the state.
+
+**A GameObject can be included as part of the state using one or more of the following methods.**
```
Comment (addisonbgross) [inline docs/integrating-with-unity/defining-states.mdx:17]: I think this line would work better as a header instead of bold, as it doesn't look like its starting a list. Something like 'Ways to include state', with each of the methods being a sub-heading.

File: docs/integrating-with-unity/defining-actions.mdx [hunk 2]
```diff
@@ -10,20 +10,20 @@ import TabItem from '@theme/TabItem';
 This section refers to an "action" as any instruction that a bot can issue to its GameObject.
 This can be anything from movement to casting abilities to navigating UI components.
 
-It is common for actions to interact with entities in the [game state](./defining-entities-and-their-states) and their corresponding GameObjects.
-The [`RGFindUtils`](../creating-bots/csharp/utilities#rgfindutils) class contains utility functions for finding such GameObjects within the scene and is
-useful for these types of actions.
+It is common for actions to interact with entities in the [game state](./defining-states) and their corresponding GameObjects.
```
Comment (vontell) [inline docs/integrating-with-unity/defining-actions.mdx:13]: Is RGFindUtils going away as well?

File: docs/integrating-with-unity/defining-actions.mdx [hunk 3]
```diff
@@ -149,27 +150,263 @@ private void AttackPlayer(GameObject player)
...
 
+:::info
+Note that by default you can only invoke actions on GameObjects for bots owned by your client id, or on global action
+handlers registered on the Regression Games overlay.
+
+_Clicking UI buttons is a global action handler available by default on the overlay.  Clicking a button is included in the example below._
 :::
 
-## Assigning Actions to GameObjects
+### `[RGAction]` Code Generation Example
```
Comment (vontell) [inline docs/integrating-with-unity/defining-actions.mdx:169]: Do we need to provide this? I think we should just allow the developer to assume this is abstracted. Maybe instead we can have this in a different page, or a link to an example of a generated class in one of our repos? I think we should find a balance of letting them know how it's done, but then not overloading them too much since they really don't need to know.

File: docs/integrating-with-unity/defining-actions.mdx [hunk 4]
```diff
@@ -149,27 +150,263 @@ private void AttackPlayer(GameObject player)
...
+                { "characterType", "Mage" }
+            };
+        }
+
+        public override void ProcessTick(RG rgObject)
+        {
+            switch (rgObject.SceneName)
+            {
+                case "CharSelect":
+                    var readyButton = rgObject.GetInteractableButtonByScenePath("/CharacterSelectCanvas/ClassInfoBox/DecorativeFrame/Ready Btn");
```
Comment (vontell) [inline docs/integrating-with-unity/defining-actions.mdx:356]: Oh this is cool

File: docs/integrating-with-unity/defining-actions.mdx [hunk 5]
```diff
@@ -149,27 +150,263 @@ private void AttackPlayer(GameObject player)
...
+```
+
+
+
+
+
+
+
+
+
```
Comment (vontell) [inline docs/integrating-with-unity/defining-actions.mdx:406]: Empty whitespace can be removed.

File: docs/integrating-with-unity/defining-states.mdx [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+## 1 - `RGStatefulGameObject` MonoBehaviour
+By adding the `RGStatefulGameObject` behaviour to a GameObject, that GameObject will be included in the state.
+Unless one of the other methods is used, this will include only the 'core' state properties of the GameObject.
```
Comment (vontell) [inline docs/integrating-with-unity/defining-states.mdx:21]: Might be good to define what the 'core' is here.

Comment (vontell) [reply docs/integrating-with-unity/defining-states.mdx:21]: I see now that is below - maybe we bring that up here or at least make the word 'core' a link to the section below.

Comment (vontell) [reply docs/integrating-with-unity/defining-states.mdx:21]: Actually, maybe disregard... in Markdown it seems far away, but any reader on the site should see the next section pretty easily as they read this.

File: docs/integrating-with-unity/defining-states.mdx [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    public int Health()
+    {
+        return 100;
+    }
+}
+```
+
+Generated File: Generated_RGStateEntity_KingState.cs
```
Comment (vontell) [inline docs/integrating-with-unity/defining-states.mdx:77]: Again, not sure if we need to include this, or maybe we provide a link to a separate page? Thinking about scanability of the page as well

File: docs/integrating-with-unity/defining-states.mdx [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+sidebar_label: 'Defining States'
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Defining States
+
+:::info
+TODO: We need to write a better description about why it is important to limit the state exposed to your bot.
```
Comment (vontell) [inline docs/integrating-with-unity/defining-states.mdx:11]: As I read through this, what comes to mind is the following (not sure how true it is yet though):
- Abstracting away state is important for keep code clean and independent from the rest of your game logic
- It's important to only include the state you need for both efficiency and to make sure bot logic only uses what it can see
- It helps collect state and information from across multiple scripts and game objects into one organized location

**Review by vontell — CHANGES_REQUESTED** [review]
Just a few small comments but not really any crazy blocking comments. Looking good, and I'm excited to try this out!

File: docs/integrating-with-unity/defining-actions.mdx (hunk 2 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-actions.mdx:13]: `RGFindUtils` still exists, but isn't necessary or useful to the developer in the context of actions.  Only the click button action uses it anymore and that is all hidden internally in a custom action implementation.

Regular actions work now because we dynamically add an `RGActionHandler` Behaviour at runtime to any GameObject that has any Behaviour(s) with [RGAction] attributes.  The action is then invoked against that actionhandler which has mappings of action names to Delegates of how to invoke that action correctly on the correct Behaviour on that same gameObject.

TLDR;  Unless you're doing something outside the box like trying to call an action on a gameobject not controlled by your bot client... The action handling is now all efficiently handled through the actionhandler, avoiding having to find objects on every action call.

Comment (vontell) [reply docs/integrating-with-unity/defining-actions.mdx:13]: I see, very cool!

File: docs/integrating-with-unity/defining-actions.mdx (hunk 3 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-actions.mdx:169]: Personally I'm against removing it.

Maybe we can improve the wording or transitions around it.

I started with just have the Bot action implementation part, but it lacked context for where `RGActionRequest_Player_MoveToPosition` came from or how it was defined.

Comment (vontell) [reply docs/integrating-with-unity/defining-actions.mdx:169]: Fine with me - I'm not super opposed and we can always use feedback and data to support my claim if my assumption is true.

File: docs/integrating-with-unity/defining-states.mdx (hunk 1 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-states.mdx]: updated

File: docs/integrating-with-unity/defining-actions.mdx (hunk 5 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-actions.mdx]: whitespace has been removed

File: docs/integrating-with-unity/defining-states.mdx (hunk 6 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-states.mdx]: all references to core are now links to the core state properties table

File: docs/integrating-with-unity/defining-states.mdx (hunk 8 above)
Comment (RG-nAmKcAz) [reply docs/integrating-with-unity/defining-states.mdx]: updated

**Review by vontell — APPROVED** [review]
